src/bootstrap/src/core/build_steps/clippy.rs

```diff
@@ -386,7 +512,21 @@ impl Step for CI {
     }
 
     fn run(self, builder: &Builder<'_>) -> Self::Output {
+        if builder.top_stage != 2 {
+            eprintln!("ERROR: `x clippy ci` should always be executed with --stage 2");
+            exit!(1);
+        }
+
+        // We want to check in-tree source using in-tree clippy. However, if we naively did
+        // a stage 2 `x clippy ci`, it would *build* a stage 2 rustc, in order to lint stage 2
+        // std, which is wasteful.
+        // So we want to lint stage 2 [bootstrap/rustc/...], but only stage 1 std rustc_codegen_gcc.
+        // We thus construct the compilers in this step manually, to optimize the number of
+        // steps that get built.
+
         builder.ensure(Bootstrap {
+            // This will be the stage 1 compiler
+            build_compiler: prepare_compiler_for_check(builder, self.target, Mode::ToolTarget),
```

Comment:
Remark: hm... this doesn't feel great (it feels a bit... *quirky*), but I don't see how we can avoid doing this if we want to maintain the core staging invariants